File: June-LeetCoding-Challenge/27-Perfect-Squares/Perfect-Squares.java
class Solution {
    public int numSquares(int n) {
        int[] dp = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            int j = 1;
            int min = Integer.MAX_VALUE;
            while (i >= j * j) {
                min = Math.min(min, dp[i - j * j] + 1);
                j++;
            }
            dp[i] = min;
        }
        return dp[n];
    }
}

File: June-LeetCoding-Challenge/27-Perfect-Squares/Perfect-Squares.cs
// Runtime: 88 ms, faster than 90.00% of C# online submissions for Perfect Squares.
// Memory Usage: 16.1 MB, less than 88.06% of C# online submissions for Perfect Squares.
public class Solution {
    public int NumSquares(int n) {
        var dp = new int[n+1];
        dp[0] = 0;
        for(int i=1; i<=n; i++){
            var j = 1;
            var min = n;
            while(i >= j*j){
                min = Math.Min(min, dp[i - j*j] +1);
                ++j;
            }
            dp[i] = min;
        }
        return dp[n];
    }
}